File: 07_Recursion/JavaScript/countchange.js
function change(money, coins) {
    return (function f(sum,coins) { return sum > money || coins.length == 0 ? 0 : sum == money ? 1 : f(coins[0] + sum, coins) + f(sum, coins.slice(1)); })(0, coins);
}

console.log(change(9,[1,2,5]));
console.log(change(4,[1,2]));

File: 07_Recursion/j4g/recursion/CountChange.scala
package j4g.recursion

/**
  * Created by nacho on 05/02/17.
  */
object CountChange extends App {
  def countChange(money:Int, coins:List[Int]): Int = {
    def count(coins: List[Int], acc: Int): Int =
      if (acc > money || coins.isEmpty) 0
      else if (acc == money) 1
      else count(coins, acc + coins.head) + count(coins.tail, acc)
    count(coins, 0)
  }

  println(countChange(4,List(1,2)))
  println(countChange(9,List(1,2,5)))
}
